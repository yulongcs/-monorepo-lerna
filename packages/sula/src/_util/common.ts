import assignWith from 'lodash/assignWith';
import isUndefined from 'lodash/isUndefined';
import isArray from 'lodash/isArray';

function withDefined(dest: any, src: any) {
  return isUndefined(src) ? dest : src;
}

export function assignWithDefined(...args: Record<string, any>[]) {
  return assignWith(...args, withDefined);
}

export function toArray<T>(value?: T | T[] | null): T[] {
  if (value === undefined || value === null) {
    return [];
  }

  // @ts-ignore
  return isArray(value) ? value : [value];
}


