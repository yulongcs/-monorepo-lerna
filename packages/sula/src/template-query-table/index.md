---
group:
  title: 模板 query-table
title: QueryTable
---

<code title="basic" src="./demos/basic.tsx" />

<code title="formaction" src="./demos/formaction.tsx" />