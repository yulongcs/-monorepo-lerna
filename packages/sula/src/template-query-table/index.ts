export type { QueryTableProps } from './QueryTable';
export type { QueryFormProps } from './QueryForm';

export { default as QueryTable } from './QueryTable';
export { default as QueryForm } from './QueryForm';
