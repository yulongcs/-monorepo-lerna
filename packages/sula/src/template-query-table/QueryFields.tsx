import React from 'react';
import DownOutlined from '@ant-design/icons/DownOutlined';
import cx from 'classnames';
import { getItemSpan } from '../form/utils/layoutUtil';
import FieldGroupContext from '../form/FieldGroupContext';
import { FieldGroup, Field, FormAction, FieldProps, FormInstance, FormProps } from '../form';
import './style/query-fields.less';
import LocaleReceiver from '../localereceiver';
import { toArray } from '../_util/common';

export interface QueryFieldsProps {
  fields: FieldProps[];
  visibleFieldsCount?: number | true;
  hasBottomBorder?: boolean;
  actionsRender: FormProps['actionsRender'],
  /** @private */
  getFormInstance: () => FormInstance;
}

export default class QueryFields extends React.Component<QueryFieldsProps> {
  static contextType = FieldGroupContext;

  static defaultProps = {
    visibleFieldsCount: 5,
    fields: [],
  };

  state = {
    collapsed: true,
  };

  getVisibleFieldsCount = (): number => {
    if (this.props.visibleFieldsCount === true) {
      return this.props.fields.length;
    }

    return this.props.visibleFieldsCount!;
  };

  renderFields = () => {
    const { fields } = this.props;

    const { matchedPoint, itemLayout } = this.context;

    const fieldsNameList = [];

    let allSpan = 0;
    let visibleAllSpan = 0;

    const visibleFieldsCount = this.getVisibleFieldsCount();

    const finalFields = fields.map((field, index) => {
      fieldsNameList.push(field.name);
      const isVisible = index < visibleFieldsCount;
      const itemSpan = getItemSpan(itemLayout, matchedPoint, field.itemLayout);
      allSpan += itemSpan;
      if (isVisible) {
        visibleAllSpan += itemSpan;
      }
      return <Field {...field} initialVisible={isVisible} key={field.name} />;
    });
    this.expandSpan = 24 - (allSpan % 24);
    this.collapseSpan = 24 - (visibleAllSpan % 24);

    return finalFields;
  };

  hasMoreQueryFields() {
    const visibleFieldsCount = this.getVisibleFieldsCount();
    return visibleFieldsCount < this.props.fields.length;
  }

  updateVisibleFields() {
    const { getFormInstance, fields } = this.props;
    const formInstance = getFormInstance();

    const visibleFieldsCount = this.getVisibleFieldsCount();

    fields.forEach((field, index) => {
      if (index >= visibleFieldsCount) {
        formInstance.setFieldVisible(field.name, !this.state.collapsed);
      }
    });
  }

  renderFormAction = (locale) => {
    const { layout } = this.context;
    const { collapsed } = this.state;
    const actionsRender = [
      ...(toArray(this.props.actionsRender)),
      ...(this.hasMoreQueryFields()
        ? [
            {
              type: () => (
                <a>
                  <span>{collapsed ? locale.expandText : locale.collapseText}</span>
                  <DownOutlined
                    style={{
                      transition: '0.3s all',
                      transform: `rotate(${collapsed ? 0 : 0.5}turn)`,
                    }}
                  />
                </a>
              ),
              action: () => {
                this.setState(
                  {
                    collapsed: !collapsed,
                  },
                  () => {
                    this.updateVisibleFields();
                  },
                );
              },
            },
          ]
        : []),
    ];

    const finalSpan = this.state.collapsed ? this.collapseSpan : this.expandSpan;
    const layoutProps = {} as any;

    if (finalSpan === 24) {
      layoutProps.actionsPosition = 'right';
      layoutProps.style = {
        marginBottom: 24,
      };
    } else {
      layoutProps.style = {
        display: 'flex',
        justifyContent: 'flex-end',
        ...(layout === 'vertical' ? { marginTop: 30 } : {}),
      };
    }

    return (
      <FormAction
        itemLayout={{
          span: finalSpan,
        }}
        {...layoutProps}
        actionsRender={actionsRender}
      />
    );
  };

  render() {
    const { hasBottomBorder } = this.props;
    return (
      <LocaleReceiver>
        {(locale) => {
          return (
            <FieldGroup
              container={{
                type: 'div',
                props: {
                  className: cx(`sula-template-query-table-fields-wrapper`, {
                    [`sula-template-query-table-fields-divider`]: hasBottomBorder,
                  }),
                },
              }}
            >
              {this.renderFields()}
              {this.renderFormAction(locale)}
            </FieldGroup>
          );
        }}
      </LocaleReceiver>
    );
  }
}
