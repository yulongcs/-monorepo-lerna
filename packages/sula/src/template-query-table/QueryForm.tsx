import React from 'react';
import { Form, FormInstance, FormProps } from '../form';
import QueryFields, { QueryFieldsProps } from './QueryFields';

export interface QueryFormProps
  extends Omit<FormProps, 'fields' | 'actionsPosition' | 'actionsRender'>,
    Omit<QueryFieldsProps, 'getFormInstance'> {}

const QueryForm: React.ForwardRefRenderFunction<FormInstance, QueryFormProps> = (props, ref) => {
  const {
    layout,
    itemLayout = {
      cols: 3,
    },
    fields,
    initialValues,
    visibleFieldsCount,
    actionsRender,
    hasBottomBorder,
    ...restProps
  } = props;
  const [form] = Form.useForm(restProps.form);

  React.useImperativeHandle(ref, () => form);

  return (
    <Form
      {...restProps}
      form={form}
      initialValues={initialValues}
      itemLayout={itemLayout}
      layout={layout}
    >
      <QueryFields
        fields={fields}
        visibleFieldsCount={visibleFieldsCount}
        getFormInstance={() => form}
        hasBottomBorder={hasBottomBorder}
        actionsRender={actionsRender}
      />
    </Form>
  );
};

export default React.forwardRef(QueryForm);
