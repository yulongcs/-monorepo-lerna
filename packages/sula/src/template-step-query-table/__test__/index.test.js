import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { delay } from '../../__tests__/common';
import { QueryTable } from '../../template-query-table';
import Memorize from '../memorize';
import { StepQueryTable } from '..';

const config = {
  rowKey: 'id',
  fields: [
    {
      name: 'name',
      label: 'name',
      field: 'input',
    },
  ],
  columns: [
    {
      key: 'id',
      title: 'Id',
    },
    {
      key: 'name',
      title: 'name',
    },
  ],
  steps: [
    {
      title: 'Steps1',
    },
    {
      title: 'Steps2',
    },
    {
      title: 'Steps3',
    },
  ],
  remoteDataSource: {
    url: '/datasource.json',
    method: 'post',
  },
};

describe('step query table', () => {
  describe('step query table', () => {
    function getMemorizeStyle(wrapper, idx = 0) {
      return wrapper.find('Memorize').children().childAt(idx).props().style;
    }

    async function changeStep(wrapper, idx) {
      wrapper.find('.ant-steps-item-container').at(idx).simulate('click');

      await act(async () => {
        await delay(1000);
        wrapper.update();
      });
    }

    function isActionStep(wrapper, idx) {
      expect(
        wrapper.find('.ant-steps-item').at(idx).hasClass('ant-steps-item-active'),
      ).toBeTruthy();
    }

    it('step', async () => {
      const wrapper = mount(<StepQueryTable {...config} />);

      expect(wrapper.find(QueryTable).length).toBe(1);

      await changeStep(wrapper, 1);
      isActionStep(wrapper, 1);
      wrapper.update();
      expect(wrapper.find(QueryTable).length).toBe(2);
      expect(getMemorizeStyle(wrapper)).toEqual({ display: 'none' });

      await changeStep(wrapper, 2);
      isActionStep(wrapper, 2);
      expect(wrapper.find(QueryTable).length).toBe(3);
      expect(getMemorizeStyle(wrapper, 1)).toEqual({ display: 'none' });
    });

    it('different step', async () => {
      const wrapper = mount(
        <StepQueryTable
          {...config}
          steps={[
            {
              title: 'Steps1',
              columns: [
                {
                  key: 'id',
                  title: 'Id',
                },
                {
                  key: 'name',
                  title: 'name',
                },
                {
                  key: 'age',
                  title: 'age',
                },
              ],
            },
            {
              title: 'Steps2',
            },
          ]}
        />,
      );

      expect(wrapper.find('.ant-table-thead').find('tr').children().length).toEqual(3);

      await changeStep(wrapper, 1);
      expect(wrapper.find('.ant-table-thead').last().find('tr').children().length).toEqual(2);
    });

    it('autoRefresh', async () => {
      let times = 0;
      const wrapper = mount(
        <StepQueryTable
          {...config}
          remoteDataSource={{
            url: '/datasource.json',
            method: 'post',
            convertParams: ({ params }) => {
              times += 1;
              return params;
            },
          }}
        />,
      );

      await changeStep(wrapper, 1);
      await changeStep(wrapper, 0);

      expect(times).toEqual(3);
    });
  });

  describe('memorize', () => {
    it('memorize', () => {
      expect(mount(<Memorize.Item />).html()).toBeNull();
    });
  });

  it('snapshot', () => {
    const wrapper = mount(<StepQueryTable {...config} />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
