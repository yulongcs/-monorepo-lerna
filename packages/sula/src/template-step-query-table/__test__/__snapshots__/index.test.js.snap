// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step query table snapshot 1`] = `
<div>
  <div
    class="sula-template-step-query-table-steps"
  >
    <div
      class="ant-steps ant-steps-horizontal ant-steps-label-horizontal"
    >
      <div
        class="ant-steps-item ant-steps-item-process ant-steps-item-active"
      >
        <div
          class="ant-steps-item-container"
          role="button"
          tabindex="0"
        >
          <div
            class="ant-steps-item-tail"
          />
          <div
            class="ant-steps-item-icon"
          >
            <span
              class="ant-steps-icon"
            >
              1
            </span>
          </div>
          <div
            class="ant-steps-item-content"
          >
            <div
              class="ant-steps-item-title"
            >
              Steps1
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-steps-item ant-steps-item-wait"
      >
        <div
          class="ant-steps-item-container"
          role="button"
          tabindex="0"
        >
          <div
            class="ant-steps-item-tail"
          />
          <div
            class="ant-steps-item-icon"
          >
            <span
              class="ant-steps-icon"
            >
              2
            </span>
          </div>
          <div
            class="ant-steps-item-content"
          >
            <div
              class="ant-steps-item-title"
            >
              Steps2
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-steps-item ant-steps-item-wait"
      >
        <div
          class="ant-steps-item-container"
          role="button"
          tabindex="0"
        >
          <div
            class="ant-steps-item-tail"
          />
          <div
            class="ant-steps-item-icon"
          >
            <span
              class="ant-steps-icon"
            >
              3
            </span>
          </div>
          <div
            class="ant-steps-item-content"
          >
            <div
              class="ant-steps-item-title"
            >
              Steps3
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div>
        <form
          class="ant-form ant-form-horizontal"
        >
          <div
            class="sula-template-query-table-fields-wrapper"
          >
            <div
              class="ant-row"
              style="margin-left: -12px; margin-right: -12px;"
            >
              <div
                class="ant-col ant-col-8"
                style="padding-left: 12px; padding-right: 12px;"
              >
                <div
                  class="ant-row ant-form-item"
                >
                  <div
                    class="ant-col ant-col-8 ant-form-item-label"
                  >
                    <label
                      class=""
                      for="name"
                      title="name"
                    >
                      name
                    </label>
                  </div>
                  <div
                    class="ant-col ant-col-16 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <input
                          class="ant-input"
                          id="name"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-col ant-col-16"
                style="padding-left: 12px; padding-right: 12px;"
              >
                <div
                  class="ant-row ant-form-item"
                  style="margin-bottom: 0px;"
                >
                  <div
                    class="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="sula-form-action-default"
                          style="display: flex; justify-content: flex-end;"
                        >
                          <div
                            class="ant-space ant-space-horizontal ant-space-align-center"
                          >
                            <div
                              class="ant-space-item"
                              style="margin-right: 8px;"
                            >
                              <button
                                class="ant-btn ant-btn-primary"
                                type="button"
                              >
                                <span>
                                  Query
                                </span>
                              </button>
                            </div>
                            <div
                              class="ant-space-item"
                            >
                              <button
                                class="ant-btn"
                                type="button"
                              >
                                <span>
                                  Reset
                                </span>
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
        <div
          class="ant-table-wrapper sula-template-query-table"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div>
              <div
                class="ant-spin ant-spin-spinning"
              >
                <span
                  class="ant-spin-dot ant-spin-dot-spin"
                >
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                  <i
                    class="ant-spin-dot-item"
                  />
                </span>
              </div>
            </div>
            <div
              class="ant-spin-container ant-spin-blur"
            >
              <div
                class="ant-table ant-table-empty"
              >
                <div
                  class="ant-table-container"
                >
                  <div
                    class="ant-table-content"
                  >
                    <table
                      style="table-layout: auto;"
                    >
                      <colgroup />
                      <thead
                        class="ant-table-thead"
                      >
                        <tr>
                          <th
                            class="ant-table-cell"
                          >
                            Id
                          </th>
                          <th
                            class="ant-table-cell"
                          >
                            name
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="ant-table-tbody"
                      >
                        <tr
                          class="ant-table-placeholder"
                        >
                          <td
                            class="ant-table-cell"
                            colspan="2"
                          >
                            <div
                              class="ant-empty ant-empty-normal"
                            >
                              <div
                                class="ant-empty-image"
                              >
                                <svg
                                  class="ant-empty-img-simple"
                                  height="41"
                                  viewBox="0 0 64 41"
                                  width="64"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <g
                                    fill="none"
                                    fill-rule="evenodd"
                                    transform="translate(0 1)"
                                  >
                                    <ellipse
                                      class="ant-empty-img-simple-ellipse"
                                      cx="32"
                                      cy="33"
                                      rx="32"
                                      ry="7"
                                    />
                                    <g
                                      class="ant-empty-img-simple-g"
                                      fill-rule="nonzero"
                                    >
                                      <path
                                        d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                      />
                                      <path
                                        class="ant-empty-img-simple-path"
                                        d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </div>
                              <p
                                class="ant-empty-description"
                              >
                                No Data
                              </p>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <form
          class="ant-form ant-form-horizontal"
        >
          <div
            class=""
          >
            <div
              class="ant-drawer ant-drawer-right"
              tabindex="-1"
            >
              <div
                class="ant-drawer-mask"
              />
              <div
                class="ant-drawer-content-wrapper"
                style="transform: translateX(100%); width: 550px;"
              >
                <div
                  class="ant-drawer-content"
                >
                  <div
                    class="ant-drawer-wrapper-body"
                  >
                    <div
                      class="ant-drawer-header-no-title"
                    >
                      <button
                        aria-label="Close"
                        class="ant-drawer-close"
                      >
                        <span
                          aria-label="close"
                          class="anticon anticon-close"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                            />
                          </svg>
                        </span>
                      </button>
                    </div>
                    <div
                      class="ant-drawer-body"
                    >
                      <div
                        class="ant-spin-nested-loading"
                      >
                        <div
                          class="ant-spin-container"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
        <form
          class="ant-form ant-form-horizontal"
        >
          <div
            class="ant-modal-root"
          >
            <div
              class="ant-modal-mask ant-modal-mask-hidden"
            />
            <div
              class="ant-modal-wrap "
              role="dialog"
              style="display: none;"
              tabindex="-1"
            >
              <div
                class="ant-modal "
                role="document"
                style="width: 520px;"
              >
                <div
                  aria-hidden="true"
                  style="width: 0px; height: 0px; overflow: hidden; outline: none;"
                  tabindex="0"
                />
                <div
                  class="ant-modal-content"
                >
                  <button
                    aria-label="Close"
                    class="ant-modal-close"
                    type="button"
                  >
                    <span
                      class="ant-modal-close-x"
                    >
                      <span
                        aria-label="close"
                        class="anticon anticon-close ant-modal-close-icon"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                  <div
                    class="ant-modal-body"
                  >
                    <div
                      class="ant-spin-nested-loading"
                    >
                      <div
                        class="ant-spin-container"
                      />
                    </div>
                  </div>
                </div>
                <div
                  aria-hidden="true"
                  style="width: 0px; height: 0px; overflow: hidden; outline: none;"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
